### PR TITLE
PR: Refactor Spyder configuration system to use an observer pattern

### PR DESCRIPTION
**Pull request by andfoy**
## Description of Changes
This PR enables the Spyder configuration system to use an observer pattern in order to propagate changes across all the application actions/options/etc. This PR will supersede the current API configuration propagation logic

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: @andfoy

### CODE REVIEW

**Comment by jitseniesen** [comment]
Thanks on tackling this area because I found the old API hard to work with. Where is the new design explained? I am lazy and don't want to search through 100+ commits and files changed. :smile:

**Comment by andfoy** [comment]
@jitseniesen Basically the idea is to register any object that requires updates on a configuration option of a section via `CONF.observe_configuration(observer, section, option)`. If the option is None, the object is notified on any change on the section. Likewise, since now we support tuple options to access nested dictionary keys, an object can subscribe to a partial tuple key. A registered object will be notified via the `on_configuration_change(option, section, value)` method each time a registered option/section changes via `CONF.set`.

This PR also introduces two decorators `spyder.api.decorators.configuration_observer` and `spyder.api.decorators.on_conf_change` to register automatically methods that receive changes on specific keys, e.g.,

```python
from spyder.config.manager import CONF
from spyder.api.decorators import configuration_observer, on_conf_change

@configuration_observer
class ConfObserver:
    def __init__(self):
        # Register class to listen for changes in all the registered options
        for section in self._configuration_listeners:
            observed_options = self._configuration_listeners[section]
            for option in observed_options:
                CONF.observe_configuration(self, section, option)

    def on_configuration_change(self, option, section, value):
        section_recievers = self._configuration_listeners.get(section, {})
        option_recievers = section_recievers.get(option, [])
        for receiver in option_recievers:
            method = getattr(self, receiver)
            method(value)

    @on_conf_change(section='editor')
    def section_change_reciever(self, section):
          # section contains the new values of the editor section
          ...
    
    @on_conf_change(section='main_interpreter', option='custom_interpreter')
    def on_interpreter_change(self, interpreter):
          # interpreter contains the value of main_interpreter
          ...

```

Furthermore, to simplify things, this PR adds a new class mixin `spyder.api.mixins.SpyderConfigurationObserver` that already implements the aforementioned protocol. By inheriting from this mixin, the previous example can be simplified to

```python
from spyder.api.mixins import SpyderConfigurationObserver

class ConfObserver(SpyderConfigurationObserver):
    @on_conf_change(section='editor')
    def section_change_reciever(self, section):
          # section contains the new values of the editor section
          ...
    
    @on_conf_change(section='main_interpreter', option='custom_interpreter')
    def on_interpreter_change(self, interpreter):
          # interpreter contains the value of main_interpreter
          ...
```

FInally, all new actions created using `self.create_action` of the new API now can specify a section and option that is read and written to, thus they are also subject to the new observer pattern, where they will change their state if a change is done on the configuration option observed.

File: spyder/api/decorators.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# -*- coding: utf-8 -*-
```
Comment (ccordoba12) [inline spyder/api/decorators.py:1]: Move to `spyder/api/config`

File: spyder/api/mixins.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
+# -*- coding: utf-8 -*-
```
Comment (ccordoba12) [inline spyder/api/mixins.py:1]: Move to `spyder/api/config`

File: spyder/api/widgets/mixins.py [hunk 3]
```diff
@@ -12,202 +12,26 @@
...
-
-    def on_option_update(self, option, value):
-        """
-        When an option is set or changed, this method is called.
-        """
-        raise NotImplementedError(
-            'Widget must define a `on_option_update` method!')
-
-
 class SpyderToolButtonMixin:
```
Comment (ccordoba12) [inline spyder/api/widgets/mixins.py:35]: Add the option kwarg to this class too.

File: spyder/app/tests/test_mainwindow.py [hunk 4]
```diff
@@ -3669,7 +3669,11 @@ def test_ordering_lsp_requests_at_startup(main_window, qtbot):
         (4, 'textDocument/didOpen'),
     ]
 
-    assert python_client['instance']._requests[:5] == expected_requests
+    lsp_requests = python_client['instance']._requests
+    start_idx = lsp_requests.index((0, 'initialize'))
+    # print(python_client['instance']._requests)
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:3674]: Remove comment.

File: spyder/config/types.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        option: ConfigurationKey
+            Configuration option that did change.
+        section: str
+            Name of the section where `option` is contained.
+        value: Any
+            New value of the configuration option that produced the event.
+        """
+        ...
+
+    def __hash__(self) -> int: ...
```
Comment (ccordoba12) [inline spyder/config/types.py:41]: Add EOL here.

File: spyder/plugins/application/container.py [hunk 6]
```diff
@@ -85,7 +85,7 @@ class ApplicationContainer(PluginMainContainer):
         'show_internal_errors': True,
     }
 
-    def setup(self, options=DEFAULT_OPTIONS):
```
Comment (ccordoba12) [inline spyder/plugins/application/container.py:88]: Don't forget to remove DEFAULT_OPTIONS and on_option_update everywhere.

File: spyder/plugins/explorer/widgets/tests/test_explorer.py [hunk 7]
```diff
@@ -196,6 +196,7 @@ def test_single_click_to_open(qtbot, file_explorer):
     qtbot.keyClick(treewidget, Qt.Key_Up)  # To focus and select the 1st item
     initial_index = treewidget.currentIndex()  # To keep a reference
 
+    # qtbot.wait(20000)
```
Comment (ccordoba12) [inline spyder/plugins/explorer/widgets/tests/test_explorer.py:199]: Remove comment.

File: spyder/plugins/findinfiles/tests/test_widgets.py [hunk 8]
```diff
@@ -55,14 +57,30 @@ def findinfiles(qtbot, request):
...
 
+    plugin_mock = MagicMock()
+    plugin_mock.CONF_SECTION = 'find_in_files'
     if param:
-        options = FindInFilesWidget.DEFAULT_OPTIONS.copy()
-        options.update(param)
-        widget = FindInFilesWidget('find_in_files', options=options)
-        widget._setup(options=options)
-        widget.setup(options=options)
+        # print(param)
```
Comment (ccordoba12) [inline spyder/plugins/findinfiles/tests/test_widgets.py:63]: Remove comment.

File: spyder/plugins/findinfiles/tests/test_widgets.py [hunk 9]
```diff
@@ -55,14 +57,30 @@ def findinfiles(qtbot, request):
...
+        prev_values = {}
+        for param_name in param:
+            value = param[param_name]
+            prev_values[param_name] = CONF.get('find_in_files', param_name)
+            CONF.set('find_in_files', param_name, value)
+
+        # options = FindInFilesWidget.DEFAULT_OPTIONS.copy()
+        # options.update(param)
```
Comment (ccordoba12) [inline spyder/plugins/findinfiles/tests/test_widgets.py:71]: Remove comment.

File: spyder/plugins/findinfiles/tests/test_widgets.py [hunk 10]
```diff
@@ -123,6 +141,11 @@ def test_find_in_files_search(findinfiles, qtbot):
     The results of the test should be equal to the expected search result
     values.
     """
+    print(CONF.items('find_in_files'))
+    CONF.set('find_in_files', 'exclude',
+             ['*.csv', '*.dat', '*.log', '*.tmp', '*.bak', '*.orig'])
+    CONF.set('find_in_files', 'path_history', [])
+    print(CONF.items('find_in_files'))
```
Comment (ccordoba12) [inline spyder/plugins/findinfiles/tests/test_widgets.py:148]: Remove comments and CONF's

File: spyder/plugins/findinfiles/tests/test_widgets.py [hunk 11]
```diff
@@ -570,7 +593,8 @@ def test_max_history(searchin_combobox, mocker):
 def test_max_results(findinfiles, qtbot):
     """Test max results correspond to expected results."""
     value = 2
-    findinfiles.change_option('max_results', value)
+    # findinfiles.set_conf('max_results', value)
```
Comment (ccordoba12) [inline spyder/plugins/findinfiles/tests/test_widgets.py:596]: Remove comment

File: spyder/utils/qthelpers.py [hunk 12]
```diff
@@ -342,6 +343,36 @@ def create_action(parent, text, shortcut=None, icon=None, tip=None,
     return action
 
 
+def setup_toggled_action(action, toggled, section, option):
```
Comment (ccordoba12) [inline spyder/utils/qthelpers.py:346]: Add docstrings.

**Review by ccordoba12 — COMMENTED** [review]
Thanks @andfoy! I left an initial review for you.

File: spyder/plugins/help/widgets.py [hunk 13]
```diff
@@ -379,32 +364,39 @@ def setup(self, options):
...
+            option='automatic_import'
         )
         self.show_source_action = self.create_action(
             name=HelpWidgetActions.ToggleShowSource,
             text=_("Show Source"),
-            toggled=lambda value: self.set_option('show_source', value),
+            toggled=True,
+            option='show_source'
         )
+        self.get_action(HelpWidgetActions.ToggleShowSource)
```
Comment (ccordoba12) [inline spyder/plugins/help/widgets.py:377]: This line doesn't do anything. Please remove it.

File: spyder/plugins/help/widgets.py [hunk 14]
```diff
@@ -473,46 +465,66 @@ def setup(self, options):
...
+        action.setIcon(icon)
+        action.setToolTip(tip)
+
+    @on_conf_change(option='automatic_import')
+    def on_automatic_import_update(self, value):
+        self.object_combo.validate_current_text()
+        if (self._last_editor_doc is None or
+                self._last_console_cb is None or
+                    self._last_editor_cb is None):
+            self.show_intro_message()
```
Comment (ccordoba12) [inline spyder/plugins/help/widgets.py:491]: Why is this change necessary? It was not like this in the previous implementation.

I can see that you added these lines in several places below. Perhaps we should move them to a method?

File: spyder/utils/qthelpers.py [hunk 15]
```diff
@@ -342,6 +343,36 @@ def create_action(parent, text, shortcut=None, icon=None, tip=None,
...
+    toggled = wrap_toggled(toggled, section, option)
+    action.toggled.connect(toggled)
+    action.setCheckable(True)
+    if section is not None and option is not None:
+        CONF.observe_configuration(action, section, option)
+        add_configuration_update(action)
+
+
+def wrap_toggled(toggled, section, option):
+    """"""
```
Comment (ccordoba12) [inline spyder/utils/qthelpers.py:356]: Add docstring.

**Review by ccordoba12 — COMMENTED** [review]
@andfoy, this is a more in-depth review for you.

File: spyder/api/plugins.py [hunk 16]
```diff
@@ -967,53 +938,7 @@ def get_plugin(self, plugin_name):
...
-                    if isinstance(options, (dict, OrderedDict)):
-                        try:
-                            config_options[option] = options[option]
-                        except Exception:
-                            pass
-
-        return config_options
-
-    def get_conf_option(self, option, default=NoDefault, section=None):
+    def get_conf(self, option, default=NoDefault, section=None):
```
[the remark this answers is not in the document]
Comment (andfoy) [reply spyder/api/plugins.py:941]: This method signature is consistent with the previous one, so it is possible to get a default value for an option without declaring its section. The suggested change, would imply that it won't be possible to to do `get_conf('opt', default)` without explicitly specifying the section. Also, the current API signature is more similar to `dict.get`

File: spyder/config/main.py [hunk 17]
```diff
@@ -285,11 +290,21 @@
...
               'max_recent_projects': 10,
-              'visible_if_project_open': True
+              'visible_if_project_open': True,
+              'single_click_to_open': False,
+              'date_column': False,
+              'kind_column': True,
+              'size_column': False,
+              'show_hidden': False,
+              'type_column': False,
+              'file_associations': {},
```
[the remark this answers is not in the document]
Comment (andfoy) [reply spyder/config/main.py:300]: It caused errors when they are not defined

File: spyder/config/main.py [hunk 18]
```diff
@@ -335,6 +351,8 @@
               }),
             ('workingdir',
              {
+              'history': [],
+              'workdir': None,
```
[the remark this answers is not in the document]
Comment (andfoy) [reply spyder/config/main.py:355]: See above

File: spyder/plugins/help/widgets.py (hunk 14 above)
Comment (andfoy) [reply spyder/plugins/help/widgets.py:491]: This change is necessary in order to display the welcome page in the Help pane due to the propagation of all the configuration options at Spyder startup. If those are not defined, then a `Documentation not available` page would appear instead. The changes do not modify the behaviour of Help at all

**Review by ccordoba12 — APPROVED** [review]
Thanks a lot @andfoy! Fantastic job with this refactoring!!

**Pull request merged**